core/io/json.cpp

```diff
@@ -340,8 +425,20 @@ Error JSON::_parse_value(Variant &value, Token &token, const CharType *p_str, in
 		value = token.value;
 		return OK;
 	} else if (token.type == TK_STRING) {
+		// The string may contain a serialized variant, try to parse it
+		VariantParser::StreamString ss;
+		ss.s = token.value;
 
-		value = token.value;
+		String v_parse_err;
+		int v_err_line;
+		Error v_err;
+
+		v_err = VariantParser::parse(&ss, value, v_parse_err, v_err_line);
+
+		if (v_err != OK || _is_compatible_value(value)) {
+			// Read the whole token value as a string instead
+			value = token.value;
```

Comment:
There could be certain performance issues having to write and parse inner JSON strings for Variants, but it might be negligible, I haven't tested this. I'd prefer perfectly working feature over speed for this personally. I was thinking that some convention could be used to encode that JSON string contains a Variant, for instance `@` reserved character could be inserted at the beginning:

```json
{
    "my_color": "@Color( 1, 1, 1, 1 )",
}
```

This could also avoid `_is_compatible_value(value)` check which I added because some strings could be falsely converted to other types. For instance, `str2var("3D")` would be parsed as a number, returning `3`. This case is covered in the test project actually.